README.md
Sanic-HTTPAuth
==============

[![Build Status](https://travis-ci.org/miguelgrinberg/Flask-HTTPAuth.png?branch=master)](https://travis-ci.org/miguelgrinberg/Flask-HTTPAuth)

This a fork of [Flask-HTTPAuth](https://github.com/miguelgrinberg/Flask-HTTPAuth) for Sanic. It is a simple extension that provides Basic and Digest HTTP authentication for Sanic routes.

Still a work in progress, contributions are welcome.

Installation
------------
The easiest way to install this is through pip.
```
pip install Sanic-HTTPAuth
```

Basic authentication example
----------------------------

```python
import hashlib
from sanic import Sanic
from sanic_httpauth import HTTPBasicAuth

app = Sanic(__name__)
auth = HTTPBasicAuth()


def hash_password(salt, password):
    salted = password + salt
    return hashlib.sha512(salted.encode("utf8")).hexdigest()


app_salt = "APP_SECRET - don't do this in production"
users = {
    "john": hash_password(app_salt, "hello"),
    "susan": hash_password(app_salt, "bye"),
}


@auth.verify_password
def verify_password(username, password):
    if username in users:
        return users.get(username) == hash_password(app_salt, password)
    return False


@app.route("/")
@auth.login_required
def index(request):
    return "Hello, %s!" % auth.username()


if __name__ == "__main__":
    app.run()
```

Note: See the [Flask-HTTPAuth documentation](http://pythonhosted.org/Flask-HTTPAuth) for more complex examples that involve password hashing and custom verification callbacks.

Digest authentication example
-----------------------------

```python
from sanic import Sanic
from sanic_httpauth import HTTPDigestAuth
from sanic_session import Session

app = Sanic(__name__)
app.config["SECRET_KEY"] = "secret key here"
auth = HTTPDigestAuth()
Session(app)

users = {"john": "hello", "susan": "bye"}


@auth.get_password
def get_pw(username):
    if username in users:
        return users.get(username)
    return None


@app.route("/")
@auth.login_required
def index(request):
    return "Hello, %s!" % auth.username()


if __name__ == "__main__":
    app.run()
```

Resources
---------

- [Flask-HTTPAuth Documentation](http://flask-httpauth.readthedocs.io/en/latest/)
- [PyPI](https://pypi.org/project/Sanic-HTTPAuth)
- [Change log](https://github.com/MihaiBalint/Sanic-HTTPAuth/blob/master/CHANGES.md)


CHANGES.md
# Flask-HTTPAuth change log

**Release 0.1.2** - 2019-06-28
- Fix @auth.login_required when applied on instance methods

**Release 0.1.1** - 2019-06-27
- Update README and examples

**Release 0.1.0** - 2019-06-27
- Replace Flask with Sanic

**Release 3.3.0** - 2019-05-19

- Use constant time string comparisons [#82](https://github.com/miguelgrinberg/Flask-HTTPAuth/issues/82) ([commit1](https://github.com/miguelgrinberg/Flask-HTTPAuth/commit/788d42ea9c4d536af628e0e7f4cb1fb84fc59a8e), [commit2](https://github.com/miguelgrinberg/Flask-HTTPAuth/commit/97f0e641a6d5eb34054de1ca255e932313d441ee)) (thanks **Brendan Long**!)
- Edited and changed the usage of JWT, because in fact the code and documentation uses JWS tokens. [#79](https://github.com/miguelgrinberg/Flask-HTTPAuth/issues/79) ([commit](https://github.com/miguelgrinberg/Flask-HTTPAuth/commit/3f743c661e281d728bd2f98af8cca000a975bb8a)) (thanks **unuseless**!)
- Documentation fix [#78](https://github.com/miguelgrinberg/Flask-HTTPAuth/issues/78) ([commit](https://github.com/miguelgrinberg/Flask-HTTPAuth/commit/c38c52326b78c91d4410f347abcd8bc49cc63ca4))
- Documentation improvements [#77](https://github.com/miguelgrinberg/Flask-HTTPAuth/issues/77) ([commit](https://github.com/miguelgrinberg/Flask-HTTPAuth/commit/ce5e5b4c9e8b748eba886ded5180e1e5d5036528))
- helper release script ([commit](https://github.com/miguelgrinberg/Flask-HTTPAuth/commit/7276d8db4b695645b01f3275addbec10418da63d))

**Release 3.2.4** - 2018-06-17

- Refactored HTTPAuth login_required [#74](https://github.com/miguelgrinberg/Flask-HTTPAuth/issues/74) ([commit](https://github.com/miguelgrinberg/Flask-HTTPAuth/commit/68ee1e7a92355ba0f3f9b48c9489a67ab762e106)) (thanks **nestedsoftware**!)
- remove incorrect references to JWT in example application [#69](https://github.com/miguelgrinberg/Flask-HTTPAuth/issues/69) ([commit](https://github.com/miguelgrinberg/Flask-HTTPAuth/commit/a310b78db2b947ab70f3fc35c1a586d822acc7ca))
- Fix typo in docs [#70](https://github.com/miguelgrinberg/Flask-HTTPAuth/issues/70) ([commit](https://github.com/miguelgrinberg/Flask-HTTPAuth/commit/b6457ae5648a50df75f3c40af4b4b3f0155fc25f)) (thanks **Grey Li**!)
- Fix documentation [#67](https://github.com/miguelgrinberg/Flask-HTTPAuth/issues/67) ([commit](https://github.com/miguelgrinberg/Flask-HTTPAuth/commit/9bd8f4b4f3574c7ef3e2fb9596bc9e9981275011)) (thanks **Eugene Rymarev**!)
- correct spelling mistake [#56](https://github.com/miguelgrinberg/Flask-HTTPAuth/issues/56) ([commit](https://github.com/miguelgrinberg/Flask-HTTPAuth/commit/f7c5bbd1b3a53080171bbdc5f1f1842f7a825f6a)) (thanks **Edward Betts**!)
- travis build fix for py36 ([commit](https://github.com/miguelgrinberg/Flask-HTTPAuth/commit/6e7f32984bda8b82200793c1b3ec44ff3df3ad2b))

**Release 3.2.3** - 2017-06-05

- Include docs and tests in pypi source tarball [#55](https://github.com/miguelgrinberg/Flask-HTTPAuth/issues/55) ([commit](https://github.com/miguelgrinberg/Flask-HTTPAuth/commit/054810ee351148b14571ba0a89ec17a543c35078)) (thanks **Chandan Kumar**!)

**Release 3.2.2** - 2017-01-30

- Validate authorization header in multi auth [#51](https://github.com/miguelgrinberg/Flask-HTTPAuth/issues/51) ([commit](https://github.com/miguelgrinberg/Flask-HTTPAuth/commit/7a895d676a1b6998f58b61a177286b62dc2872f5))
- index.rst: Add a missing variable in a code snippet [#49](https://github.com/miguelgrinberg/Flask-HTTPAuth/issues/49) ([commit](https://github.com/miguelgrinberg/Flask-HTTPAuth/commit/f7fe976bbdc699e8bafaed729dfdd74d2b27d7db)) (thanks **Baptiste Fontaine**!)

**Release 3.2.1** - 2016-09-04

- add `__version__` to package ([commit](https://github.com/miguelgrinberg/Flask-HTTPAuth/commit/d188450987f226568fe0cdee0b6d480b375af64a))
- Add readme and license files to the built package [#45](https://github.com/miguelgrinberg/Flask-HTTPAuth/issues/45) ([commit](https://github.com/miguelgrinberg/Flask-HTTPAuth/commit/1c35bec606f147bb23725d6ff3b0411f06828492))

**Release 3.2.0** - 2016-08-20

- Fix TCP Connection reset by peer error [#39](https://github.com/miguelgrinberg/Flask-HTTPAuth/issues/39) ([commit](https://github.com/miguelgrinberg/Flask-HTTPAuth/commit/94f6c6d5a4866a43ff4f269eb351dce6232791a2)) (thanks **Joe Kemp**!)

**Release 3.1.2** - 2016-04-21

- Add robustness to password check ([commit](https://github.com/miguelgrinberg/Flask-HTTPAuth/commit/051fd88ee36a21a13255b4ec69e172c9ae4ad46d))

**Release 3.1.1** - 2016-03-24

- pass params to view function in MultiAuth [#36](https://github.com/miguelgrinberg/Flask-HTTPAuth/issues/36) ([commit](https://github.com/miguelgrinberg/Flask-HTTPAuth/commit/319974602e55529006b9a8a4fde04ef08e042e83)) (thanks **vovanz**!)
- add examples to flake8 build ([commit](https://github.com/miguelgrinberg/Flask-HTTPAuth/commit/61b1b71b3b29f2936ac6a2077883da1faeaad09f))
- Added multi auth tests ([commit](https://github.com/miguelgrinberg/Flask-HTTPAuth/commit/c443e7ebcc227fd3690c2cf943d414087d7b931d))
- removed dead code ([commit](https://github.com/miguelgrinberg/Flask-HTTPAuth/commit/4d2232e2a77f5e10e1731936f4ac64439049b220))

**Release 3.1.0** - 2016-03-13

- examples ([commit](https://github.com/miguelgrinberg/Flask-HTTPAuth/commit/609806a1c10264818e08ba0ce9b7babeaf101656))
- Added support for multiple authentication methods ([commit](https://github.com/miguelgrinberg/Flask-HTTPAuth/commit/6c3f94d9eda85b78a8c36cd5e05d6d9836bee2d0))
- Added change log ([commit](https://github.com/miguelgrinberg/Flask-HTTPAuth/commit/8b427b962114a6ef13badaf8f2f1b396c540955a))
- Add additional token auth test ([commit](https://github.com/miguelgrinberg/Flask-HTTPAuth/commit/29edb1948f086babbd1a9e0c87a0a35c05f0a63b))

**Release 3.0.2** - 2016-03-12

- Let callback decide what to do when authentication type does not match ([commit](https://github.com/miguelgrinberg/Flask-HTTPAuth/commit/b942f980970d2e387a80f68de4ea2bb8728b149c))

**Release 3.0.1** - 2016-03-09


tests/test_token.py
import unittest
from sanic import Sanic
from sanic_httpauth import HTTPTokenAuth
from sanic_cors import CORS
from sanic.response import text


class HTTPAuthTestCase(unittest.TestCase):
    def setUp(self):
        app = Sanic(__name__)
        app.config["SECRET_KEY"] = "my secret"
        app.config["CORS_AUTOMATIC_OPTIONS"] = True

        CORS(app)
        token_auth = HTTPTokenAuth("MyToken")

        @token_auth.verify_token
        def verify_token(token):
            return token == "this-is-the-token!"

        @token_auth.error_handler
        def error_handler(request):
            return text(
                "error", status=401, headers={"WWW-Authenticate": 'MyToken realm="Foo"'}
            )

        @app.route("/")
        def index(request):
            return text("index")

        @app.route("/protected")
        @token_auth.login_required
        def token_auth_route(request):
            return text("token_auth")

        self.app = app
        self.token_auth = token_auth
        self.client = app.test_client

    def test_token_auth_prompt(self):
        rq, response = self.client.get("/protected")
        self.assertEqual(response.status, 401)
        self.assertTrue("WWW-Authenticate" in response.headers)
        self.assertEqual(response.headers["WWW-Authenticate"], 'MyToken realm="Foo"')

    def test_token_auth_ignore_options(self):
        rq, response = self.client.options("/protected")
        self.assertEqual(response.status, 200)
        self.assertTrue("WWW-Authenticate" not in response.headers)

    def test_token_auth_login_valid(self):
        rq, response = self.client.get(
            "/protected", headers={"Authorization": "MyToken this-is-the-token!"}
        )
        self.assertEqual(response.content.decode("utf-8"), "token_auth")

    def test_token_auth_login_valid_different_case(self):
        rq, response = self.client.get(
            "/protected", headers={"Authorization": "mytoken this-is-the-token!"}
        )
        self.assertEqual(response.content.decode("utf-8"), "token_auth")

    def test_token_auth_login_invalid_token(self):
        rq, response = self.client.get(
            "/protected", headers={"Authorization": "MyToken this-is-not-the-token!"}
        )
        self.assertEqual(response.status, 401)
        self.assertTrue("WWW-Authenticate" in response.headers)
        self.assertEqual(response.headers["WWW-Authenticate"], 'MyToken realm="Foo"')

    def test_token_auth_login_invalid_scheme(self):
        rq, response = self.client.get(
            "/protected", headers={"Authorization": "Foo this-is-the-token!"}
        )
        self.assertEqual(response.status, 401)
        self.assertTrue("WWW-Authenticate" in response.headers)
        self.assertEqual(response.headers["WWW-Authenticate"], 'MyToken realm="Foo"')

    def test_token_auth_login_invalid_header(self):
        rq, response = self.client.get(
            "/protected", headers={"Authorization": "this-is-a-bad-header"}
        )
        self.assertEqual(response.status, 401)
        self.assertTrue("WWW-Authenticate" in response.headers)
        self.assertEqual(response.headers["WWW-Authenticate"], 'MyToken realm="Foo"')

    def test_token_auth_login_invalid_no_callback(self):
        token_auth2 = HTTPTokenAuth("Token", realm="foo")

        @self.app.route("/protected2")
        @token_auth2.login_required
        def token_auth_route2(request):
            return text("token_auth2")

        rq, response = self.client.get(
            "/protected2", headers={"Authorization": "Token this-is-the-token!"}
        )
        self.assertEqual(response.status, 401)
        self.assertTrue("WWW-Authenticate" in response.headers)
        self.assertEqual(response.headers["WWW-Authenticate"], 'Token realm="foo"')


tests/test_digest_get_password.py
import unittest
import re
from hashlib import md5 as basic_md5
from sanic import Sanic
from sanic_cors import CORS
from sanic_session import Session
from sanic_httpauth import HTTPDigestAuth
from sanic.response import text

from sanic_httpauth_compat import parse_dict_header


def md5(str):
    if type(str).__name__ == "str":
        str = str.encode("utf-8")
    return basic_md5(str)


def get_ha1(user, pw, realm):
    a1 = user + ":" + realm + ":" + pw
    return md5(a1).hexdigest()


class HTTPAuthTestCase(unittest.TestCase):
    def setUp(self):
        app = Sanic(__name__)
        app.config["SECRET_KEY"] = "my secret"
        app.config["CORS_AUTOMATIC_OPTIONS"] = True

        CORS(app)
        Session(app)
        digest_auth = HTTPDigestAuth()

        @digest_auth.get_password
        def get_digest_password_2(username):
            if username == "susan":
                return "hello"
            elif username == "john":
                return "bye"
            else:
                return None

        @app.route("/")
        def index(request):
            return text("index")

        @app.route("/digest")
        @digest_auth.login_required
        def digest_auth_route(request):
            return text(f"digest_auth:{digest_auth.username(request)}")

        self.app = app
        self.digest_auth = digest_auth
        self.client = app.test_client

    def test_digest_auth_prompt(self):
        req, response = self.client.get("/digest")
        self.assertEqual(response.status_code, 401)
        self.assertTrue("WWW-Authenticate" in response.headers)
        self.assertTrue(
            re.match(
                r'^Digest realm="Authentication Required",'
                r'nonce="[0-9a-f]+",opaque="[0-9a-f]+"$',
                response.headers["WWW-Authenticate"],
            )
        )

    def test_digest_auth_ignore_options(self):
        req, response = self.client.options("/digest")
        self.assertEqual(response.status_code, 200)
        self.assertTrue("WWW-Authenticate" not in response.headers)

    def test_digest_auth_login_valid(self):
        req, response = self.client.get("/digest")
        self.assertTrue(response.status_code == 401)
        header = response.headers.get("WWW-Authenticate")
        auth_type, auth_info = header.split(None, 1)
        d = parse_dict_header(auth_info)

        a1 = "john:" + d["realm"] + ":bye"
        ha1 = md5(a1).hexdigest()
        a2 = "GET:/digest"
        ha2 = md5(a2).hexdigest()
        a3 = ha1 + ":" + d["nonce"] + ":" + ha2
        auth_response = md5(a3).hexdigest()

        req, response = self.client.get(
            "/digest",
            headers={
                "Authorization": 'Digest username="john",realm="{0}",'
                'nonce="{1}",uri="/digest",response="{2}",'
                'opaque="{3}"'.format(
                    d["realm"], d["nonce"], auth_response, d["opaque"]
                )
            },
        )
        self.assertEqual(response.content, b"digest_auth:john")

    def test_digest_auth_login_bad_realm(self):
        req, response = self.client.get("/digest")
        self.assertTrue(response.status_code == 401)
        header = response.headers.get("WWW-Authenticate")
        auth_type, auth_info = header.split(None, 1)
        d = parse_dict_header(auth_info)

        a1 = "john:" + "Wrong Realm" + ":bye"
        ha1 = md5(a1).hexdigest()
        a2 = "GET:/digest"
        ha2 = md5(a2).hexdigest()
        a3 = ha1 + ":" + d["nonce"] + ":" + ha2
        auth_response = md5(a3).hexdigest()

        req, response = self.client.get(
            "/digest",
            headers={
                "Authorization": 'Digest username="john",realm="{0}",'
                'nonce="{1}",uri="/digest",response="{2}",'
                'opaque="{3}"'.format(
                    d["realm"], d["nonce"], auth_response, d["opaque"]
                )
            },
        )
        self.assertEqual(response.status_code, 401)
        self.assertTrue("WWW-Authenticate" in response.headers)
        self.assertTrue(
            re.match(
                r'^Digest realm="Authentication Required",'
                r'nonce="[0-9a-f]+",opaque="[0-9a-f]+"$',
                response.headers["WWW-Authenticate"],
            )
        )

    def test_digest_auth_login_invalid2(self):
        req, response = self.client.get("/digest")
        self.assertEqual(response.status_code, 401)
        header = response.headers.get("WWW-Authenticate")
        auth_type, auth_info = header.split(None, 1)
        d = parse_dict_header(auth_info)

        a1 = "david:" + "Authentication Required" + ":bye"
        ha1 = md5(a1).hexdigest()
        a2 = "GET:/digest"
        ha2 = md5(a2).hexdigest()
        a3 = ha1 + ":" + d["nonce"] + ":" + ha2
        auth_response = md5(a3).hexdigest()

        req, response = self.client.get(
            "/digest",
            headers={
                "Authorization": 'Digest username="david",realm="{0}",'
                'nonce="{1}",uri="/digest",response="{2}",'
                'opaque="{3}"'.format(
                    d["realm"], d["nonce"], auth_response, d["opaque"]
                )
            },
        )
        self.assertEqual(response.status_code, 401)
        self.assertTrue("WWW-Authenticate" in response.headers)
        self.assertTrue(
            re.match(
                r'^Digest realm="Authentication Required",'
                r'nonce="[0-9a-f]+",opaque="[0-9a-f]+"$',
                response.headers["WWW-Authenticate"],
            )
        )

    def test_digest_generate_ha1(self):
        ha1 = self.digest_auth.generate_ha1("pawel", "test")
        ha1_expected = get_ha1("pawel", "test", self.digest_auth.realm)
        self.assertEqual(ha1, ha1_expected)

    def test_digest_custom_nonce_checker(self):
        @self.digest_auth.generate_nonce
        def noncemaker(request):
            return "not a good nonce"

        @self.digest_auth.generate_opaque
        def opaquemaker(request):
            return "some opaque"

        verify_nonce_called = []

        @self.digest_auth.verify_nonce
        def verify_nonce(request, provided_nonce):
            verify_nonce_called.append(provided_nonce)
            return True

        verify_opaque_called = []

        @self.digest_auth.verify_opaque
        def verify_opaque(request, provided_opaque):
            verify_opaque_called.append(provided_opaque)
            return True

        req, response = self.client.get("/digest")
        self.assertEqual(response.status_code, 401)
        header = response.headers.get("WWW-Authenticate")
        auth_type, auth_info = header.split(None, 1)
        d = parse_dict_header(auth_info)

        self.assertEqual(d["nonce"], "not a good nonce")
        self.assertEqual(d["opaque"], "some opaque")

        a1 = "john:" + d["realm"] + ":bye"
        ha1 = md5(a1).hexdigest()
        a2 = "GET:/digest"
        ha2 = md5(a2).hexdigest()
        a3 = ha1 + ":" + d["nonce"] + ":" + ha2
        auth_response = md5(a3).hexdigest()

        req, response = self.client.get(
            "/digest",
            headers={
                "Authorization": 'Digest username="john",realm="{0}",'
                'nonce="{1}",uri="/digest",response="{2}",'
                'opaque="{3}"'.format(
                    d["realm"], d["nonce"], auth_response, d["opaque"]
                )
            },
        )
        self.assertEqual(response.content, b"digest_auth:john")
        self.assertEqual(
            verify_nonce_called, ["not a good nonce"], "Should have verified the nonce."
        )
        self.assertEqual(
            verify_opaque_called, ["some opaque"], "Should have verified the opaque."
        )


tests/test_digest_ha1_password.py
import unittest
from hashlib import md5 as basic_md5
from sanic import Sanic
from sanic_session import Session
from sanic_httpauth import HTTPDigestAuth
from sanic.response import text

from sanic_httpauth_compat import parse_dict_header


def md5(str):
    if type(str).__name__ == "str":
        str = str.encode("utf-8")
    return basic_md5(str)


def get_ha1(user, pw, realm):
    a1 = user + ":" + realm + ":" + pw
    return md5(a1).hexdigest()


class HTTPAuthTestCase(unittest.TestCase):
    def setUp(self):
        app = Sanic(__name__)
        app.config["SECRET_KEY"] = "my secret"

        Session(app)
        digest_auth_ha1_pw = HTTPDigestAuth(use_ha1_pw=True)

        @digest_auth_ha1_pw.get_password
        def get_digest_password(username):
            if username == "susan":
                return get_ha1(username, "hello", digest_auth_ha1_pw.realm)
            elif username == "john":
                return get_ha1(username, "bye", digest_auth_ha1_pw.realm)
            else:
                return None

        @app.route("/")
        def index(request):
            return "index"

        @app.route("/digest_ha1_pw")
        @digest_auth_ha1_pw.login_required
        def digest_auth_ha1_pw_route(request):
            return text(f"digest_auth_ha1_pw:{digest_auth_ha1_pw.username(request)}")

        self.app = app
        self.client = app.test_client

    def test_digest_ha1_pw_auth_login_valid(self):
        req, response = self.client.get("/digest_ha1_pw")
        self.assertTrue(response.status_code == 401)
        header = response.headers.get("WWW-Authenticate")
        auth_type, auth_info = header.split(None, 1)
        d = parse_dict_header(auth_info)

        a1 = "john:" + d["realm"] + ":bye"
        ha1 = md5(a1).hexdigest()
        a2 = "GET:/digest_ha1_pw"
        ha2 = md5(a2).hexdigest()
        a3 = ha1 + ":" + d["nonce"] + ":" + ha2
        auth_response = md5(a3).hexdigest()

        req, response = self.client.get(
            "/digest_ha1_pw",
            headers={
                "Authorization": 'Digest username="john",realm="{0}",'
                'nonce="{1}",uri="/digest_ha1_pw",'
                'response="{2}",'
                'opaque="{3}"'.format(
                    d["realm"], d["nonce"], auth_response, d["opaque"]
                )
            },
        )
        self.assertEqual(response.content, b"digest_auth_ha1_pw:john")


examples/basic_auth.py
#!/usr/bin/env python
"""Basic authentication example

This example demonstrates how to protect Sanic endpoints with basic
authentication, using passwords (hashed but with an unsecure app-global salt).

After running this example, visit http://localhost:5000 in your browser. To
gain access, you can use (username=john, password=hello) or
(username=susan, password=bye).
"""
import hashlib
from sanic import Sanic
from sanic_httpauth import HTTPBasicAuth

app = Sanic(__name__)
auth = HTTPBasicAuth()


def hash_password(salt, password):
    salted = password + salt
    return hashlib.sha512(salted.encode("utf8")).hexdigest()


app_salt = "APP_SECRET - don't do this in production"
users = {
    "john": hash_password(app_salt, "hello"),
    "susan": hash_password(app_salt, "bye"),
}


@auth.verify_password
def verify_password(username, password):
    if username in users:
        return users.get(username) == hash_password(app_salt, password)
    return False


@app.route("/")
@auth.login_required
def index(request):
    return "Hello, %s!" % auth.username(request)


if __name__ == "__main__":
    app.run(debug=True, host="0.0.0.0")


examples/multi_auth.py
#!/usr/bin/env python
"""Multiple authentication example

This example demonstrates how to combine two authentication methods using the
"MultiAuth" class.

The root URL for this application can be accessed via basic auth, providing
username and password, or via token auth, providing a bearer JWS token.
"""
import hashlib
from sanic import Sanic
from sanic_httpauth import HTTPBasicAuth, HTTPTokenAuth, MultiAuth
from itsdangerous import TimedJSONWebSignatureSerializer as JWS


app = Sanic(__name__)
app.config["SECRET_KEY"] = "top secret!"
jws = JWS(app.config["SECRET_KEY"], expires_in=3600)

basic_auth = HTTPBasicAuth()
token_auth = HTTPTokenAuth("Bearer")
multi_auth = MultiAuth(basic_auth, token_auth)


def hash_password(salt, password):
    salted = password + salt
    return hashlib.sha512(salted.encode("utf8")).hexdigest()


app_salt = "APP_SECRET - don't do this in production"
users = {
    "john": hash_password(app_salt, "hello"),
    "susan": hash_password(app_salt, "bye"),
}

for user in users.keys():
    token = jws.dumps({"username": user})
    print("*** token for {}: {}\n".format(user, token))


@basic_auth.verify_password
def verify_password(username, password):
    if username in users:
        return users.get(username) == hash_password(app_salt, password)
    return False


@token_auth.verify_token
def verify_token(token):
    try:
        return "username" in token_serializer.loads(token)
    except:  # noqa: E722
        return False


@app.route("/")
@multi_auth.login_required
def index(request):
    username = basic_auth.username(request)
    if not username:
        data = token_serializer.loads(token_auth.token(request))
        username = data["username"]

    return f"Hello, {username}!"


if __name__ == "__main__":
    app.run()


sanic_httpauth.py
"""
flask_httpauth
==================

This module provides Basic and Digest HTTP authentication for Flask routes.

:copyright: (C) 2014 by Miguel Grinberg.
:license:   MIT, see LICENSE for more details.
"""
import logging

from functools import wraps
from hashlib import md5
from random import Random, SystemRandom
from sanic.response import text

from sanic_httpauth_compat import safe_str_cmp, Authorization
from sanic_httpauth_compat import parse_authorization_header

__version__ = "3.3.1dev"
log = logging.getLogger(__name__)


class HTTPAuth(object):
    def __init__(self, scheme=None, realm=None):
        self.scheme = scheme
        self.realm = realm or "Authentication Required"
        self.get_password_callback = None
        self.auth_error_callback = None

        def default_get_password(username):
            return None

        def default_auth_error(request):
            return text("Unauthorized Access", status=401)

        self.get_password(default_get_password)
        self.error_handler(default_auth_error)

    def get_password(self, f):
        self.get_password_callback = f
        return f

    def error_handler(self, f):
        @wraps(f)
        def decorated(request, *args, **kwargs):
            res = f(request, *args, **kwargs)

            if res.status == 200:
                # if user didn't set status code, use 401
                res.status = 401
            if "WWW-Authenticate" not in res.headers.keys():
                res.headers["WWW-Authenticate"] = self.authenticate_header(request)
            return res

        self.auth_error_callback = decorated
        return decorated

    def authenticate_header(self, request):
        return '{0} realm="{1}"'.format(self.scheme, self.realm)

    def get_auth(self, request):
        auth = parse_authorization_header(request.headers.get("Authorization"))
        try:
            if auth is None and "Authorization" in request.headers:
                auth_type, value = request.headers["Authorization"].split(None, 1)
                auth = Authorization(auth_type, {"token": value})
        except ValueError:
            # The Authorization header is either empty or has no token
            pass

        # if the auth type does not match, we act as if there is no auth
        # this is better than failing directly, as it allows the callback
        # to handle special cases, like supporting multiple auth types
        if auth is not None and auth.type.lower() != self.scheme.lower():
            auth = None

        return auth

    def get_auth_password(self, auth):
        password = None

        if auth and auth.username:
            password = self.get_password_callback(auth.username)

        return password

    def login_required(self, f):
        @wraps(f)
        def decorated(request, *args, **kwargs):
            auth = self.get_auth(request)
            request["authorization"] = auth

            # Sanic-CORS normally handles OPTIONS requests on its own, but in the
            # case it is configured to forward those to the application, we
            # need to ignore authentication headers and let the request through
            # to avoid unwanted interactions with CORS.
            if request.method != "OPTIONS":  # pragma: no cover
                password = self.get_auth_password(auth)

                if not self.authenticate(request, auth, password):
                    return self.auth_error_callback(request)

            return f(request, *args, **kwargs)

        return decorated

    def username(self, request):
        if not request["authorization"]:
            return ""
        return request["authorization"].username


class HTTPBasicAuth(HTTPAuth):
    def __init__(self, scheme=None, realm=None):
        super(HTTPBasicAuth, self).__init__(scheme or "Basic", realm)

        self.hash_password_callback = None
        self.verify_password_callback = None

    def hash_password(self, f):
        self.hash_password_callback = f
        return f

    def verify_password(self, f):
        self.verify_password_callback = f
        return f

    def authenticate(self, request, auth, stored_password):
        if auth:
            username = auth.username
            client_password = auth.password
        else:
            username = ""
            client_password = ""
        if self.verify_password_callback:
            return self.verify_password_callback(username, client_password)
        if not auth:
            return False
        if self.hash_password_callback:
            try:
                client_password = self.hash_password_callback(client_password)
            except TypeError:
                client_password = self.hash_password_callback(username, client_password)
        return (
            client_password is not None
            and stored_password is not None
            and safe_str_cmp(client_password, stored_password)
        )


class HTTPDigestAuth(HTTPAuth):
    def __init__(self, scheme=None, realm=None, use_ha1_pw=False):
        super(HTTPDigestAuth, self).__init__(scheme or "Digest", realm)
        self.use_ha1_pw = use_ha1_pw
        self.random = SystemRandom()
        try:
            self.random.random()
        except NotImplementedError:  # pragma: no cover
            self.random = Random()

        self.generate_nonce_callback = None
        self.verify_nonce_callback = None
        self.generate_opaque_callback = None
        self.verify_opaque_callback = None

        def _generate_random():
            return md5(str(self.random.random()).encode("utf-8")).hexdigest()

        def default_generate_nonce(request):
            request["session"]["auth_nonce"] = _generate_random()
            return request["session"]["auth_nonce"]

        def default_verify_nonce(request, nonce):
            session_nonce = request["session"].get("auth_nonce")
            if nonce is None or session_nonce is None:
                return False
            return safe_str_cmp(nonce, session_nonce)

        def default_generate_opaque(request):
            request["session"]["auth_opaque"] = _generate_random()
            return request["session"]["auth_opaque"]

        def default_verify_opaque(request, opaque):
            session_opaque = request["session"].get("auth_opaque")
            if opaque is None or session_opaque is None:
                return False
            return safe_str_cmp(opaque, session_opaque)

        self.generate_nonce(default_generate_nonce)
        self.generate_opaque(default_generate_opaque)
        self.verify_nonce(default_verify_nonce)
        self.verify_opaque(default_verify_opaque)

    def generate_nonce(self, f):
        self.generate_nonce_callback = f
        return f

    def verify_nonce(self, f):
        self.verify_nonce_callback = f
        return f

    def generate_opaque(self, f):
        self.generate_opaque_callback = f
        return f

    def verify_opaque(self, f):
        self.verify_opaque_callback = f
        return f

    def get_nonce(self, request):
        return self.generate_nonce_callback(request)

    def get_opaque(self, request):
        return self.generate_opaque_callback(request)

    def generate_ha1(self, username, password):
        a1 = username + ":" + self.realm + ":" + password
        a1 = a1.encode("utf-8")
        return md5(a1).hexdigest()

    def authenticate_header(self, request):
        nonce = self.get_nonce(request)
        opaque = self.get_opaque(request)
        return '{0} realm="{1}",nonce="{2}",opaque="{3}"'.format(
            self.scheme, self.realm, nonce, opaque
        )

    def authenticate(self, request, auth, stored_password_or_ha1):
        if (
            not auth
            or not auth.username
            or not auth.realm
            or not auth.uri
            or not auth.nonce
            or not auth.response
            or not stored_password_or_ha1
        ):
            return False
        if not (self.verify_nonce_callback(request, auth.nonce)) or not (
            self.verify_opaque_callback(request, auth.opaque)
        ):
            return False
        if self.use_ha1_pw:
            ha1 = stored_password_or_ha1
        else:
            a1 = auth.username + ":" + auth.realm + ":" + stored_password_or_ha1
            ha1 = md5(a1.encode("utf-8")).hexdigest()
        a2 = request.method + ":" + auth.uri
        ha2 = md5(a2.encode("utf-8")).hexdigest()
        a3 = ha1 + ":" + auth.nonce + ":" + ha2
        response = md5(a3.encode("utf-8")).hexdigest()
        return safe_str_cmp(response, auth.response)


class HTTPTokenAuth(HTTPAuth):
    def __init__(self, scheme="Bearer", realm=None):
        super(HTTPTokenAuth, self).__init__(scheme, realm)

        self.verify_token_callback = None

    def verify_token(self, f):
        self.verify_token_callback = f
        return f

    def authenticate(self, request, auth, stored_password):
        if auth:
            token = auth["token"]
        else:
            token = ""
        if self.verify_token_callback:
            return self.verify_token_callback(token)
        return False


class MultiAuth(object):
    def __init__(self, main_auth, *args):
        self.main_auth = main_auth
        self.additional_auth = args

    def login_required(self, f):
        @wraps(f)
        def decorated(request, *args, **kwargs):
            selected_auth = None
            if "Authorization" in request.headers:
                try:
                    scheme, creds = request.headers["Authorization"].split(None, 1)
                except ValueError:
                    # malformed Authorization header
                    pass
                else:
                    for auth in self.additional_auth:
                        if auth.scheme == scheme:
                            selected_auth = auth
                            break
            if selected_auth is None:
                selected_auth = self.main_auth

            return selected_auth.login_required(f)(request, *args, **kwargs)

        return decorated
